data_structures.md

```diff
@@ -11,12 +11,601 @@ kernelspec:
 
 ## Tables
 
+Internally, a tree sequence can be thought of as a set of tables, and {program}`tskit`
+provides an interface for dealing with these tables directly. This is particularly
+relevant when you wish to {ref}`edit or otherwise modify<sec_data_structures_editing>`
+a tree sequence.
+
+There are eight tables that together define a tree sequence, although some may be empty,
+and together they form a {class}`TableCollection`.
+The tables are defined in the official {program}`tskit` documentation for
+{ref}`Table Definitions <tskit:sec_table_definitions>`, and the
+{ref}`Tables API <tskit:sec_tables_api>` section in the docs describes how to work with
+them. Here we make some general remarks about what you can and cannot do with them.
+
+{program}`tskit` provides direct access to the columns of each table as
+{program}`numpy` arrays: for instance, if ``n`` is a {class}`NodeTable`, then ``n.time``
+will return an array containing the birth times of the individuals whose genomes
+are represented by the nodes in the table.
+*However*, it is important to note that this is *not* a shallow copy:
+modifying ``n.time`` will *not* change the node table ``n``.  This may change in
+the future, but currently there are three ways to modify tables: ``.add_row()``,
+``.set_columns()``, and ``.append_columns()``
+(and also ``.clear()``, which empties the table).
+
+:::{todo}
+Update this to reflect the new slice methods for accessing and setting rows.
+:::
+
+For example, a node table could be constructed using {meth}`NodeTable.add_row` as
+follows:
+
+```{code-cell} ipython3
+import tskit
+
+n = tskit.NodeTable()
+sv = [True, True, True, False, False, False, False]
+tv = [0.0, 0.0, 0.0, 0.4, 0.5, 0.7, 1.0]
+pv = [0, 0, 0, 0, 0, 0, 0]
+for s, t, p in zip(sv, tv, pv):
+    n.add_row(flags=s, population=p, time=t)
+n
+```
+
+
+The ``.add_row()`` method is natural (and should be reasonably efficient) if
+new records appear one-by-one. In the example above it would have been more
+natural to use ``.set_columns()`` --- equivalently:
+
+```{code-cell} ipython3
+n = tskit.NodeTable()
+n.set_columns(flags=sv, population=pv, time=tv)
+```
+
+Since columns cannot be modified directly as properties of the tables,
+they must be extracted, modified, then replaced.
+:::{todo}
+The above is no longer strictly true with the slice setting methods.
+:::
+For example, here we add 1.4 to every ``time`` except the first
+in the node table constructed above (using ``numpy`` indexing):
+
+```{code-cell} ipython3
+tn = n.time
+tn[1:] = tn[1:] + 1.4
+n.set_columns(flags=n.flags, population=n.population, time=tn)
+n
+```
+
+
+### Overview of the Tables Format
+
+The {ref}`Table Definitions <tskit:sec_table_definitions>` section of the official
+{program}`tskit` documentation gives a precise
+definition of how a tree sequence is stored in a collection of tables.
+Here we give an overview. Consider the following sequence of trees:
+
+```{code-cell} ipython3
```

Comment:
I don't think you need the ``ipython3``s on the code-cells

Reply:
I was copying https://jupytext.readthedocs.io/en/latest/formats.html#myst-markdown - there it says it's "optional" but useful as an aide to syntax highlighting?